File: go/transport/rest/adapt.go
package rest

import (
	"net/http"

	"github.com/gt-tech-ai/knowledge-engine/go/transport"
)

// Adapt bridges an http.HandlerFunc to a typed transport.HandlerFunc.
//
// The adapter handles the full HTTP lifecycle:
//  1. Parse the HTTP request into a typed Req using the provided parse function
//  2. Call the decorated HandlerFunc with the parsed request
//  3. Write the response using BaseController (WriteSuccess for 200, WriteCreated for 201, etc.)
//
// Use writeStatus to control the success HTTP status code (e.g., http.StatusOK,
// http.StatusCreated). For 204 No Content responses, use AdaptNoContent instead.
//
// Example:
//
//	mux.HandleFunc("GET /api/v1/users/{id}", rest.Adapt(
//	    bc, getHandler, http.StatusOK,
//	    func(r *http.Request) (GetUserReq, error) {
//	        return GetUserReq{ID: r.PathValue("id")}, nil
//	    },
//	))
func Adapt[Req, Resp any](
	bc *BaseController,
	handler transport.HandlerFunc[Req, Resp],
	writeStatus int,
	parse func(r *http.Request) (Req, error),
) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		req, err := parse(r)
		if err != nil {
			bc.WriteError(w, err)
			return
		}

		resp, err := handler(r.Context(), req)
		if err != nil {
			bc.WriteError(w, err)
			return
		}

		bc.WriteJSON(w, writeStatus, resp)
	}
}

// AdaptNoContent bridges an http.HandlerFunc to a typed transport.HandlerFunc
// that returns 204 No Content on success (ignoring the response value).
//
// Example:
//
//	mux.HandleFunc("DELETE /api/v1/users/{id}", rest.AdaptNoContent(
//	    bc, deleteHandler,
//	    func(r *http.Request) (DeleteUserReq, error) {
//	        return DeleteUserReq{ID: r.PathValue("id")}, nil
//	    },
//	))
func AdaptNoContent[Req, Resp any](
	bc *BaseController,
	handler transport.HandlerFunc[Req, Resp],
	parse func(r *http.Request) (Req, error),
) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		req, err := parse(r)
		if err != nil {
			bc.WriteError(w, err)
			return
		}

		_, err = handler(r.Context(), req)
		if err != nil {
			bc.WriteError(w, err)
			return
		}

		bc.WriteNoContent(w)
	}
}

File: python/techai_webutils/src/techai_webutils/controllers/adapt.py
"""REST adapter bridging typed handlers to framework handlers.

Mirrors Go's ``go/transport/rest/adapt.go`` ``Adapt[Req, Resp]()`` and
``AdaptNoContent[Req, Resp]()``.

Provides a generic bridge from typed ``HandlerFunc`` to framework request
handlers (e.g., FastAPI, Starlette). The parse function extracts the typed
request from the raw HTTP request; the handler processes it; the response
is serialized back.
"""

from __future__ import annotations

from typing import Any, TYPE_CHECKING


if TYPE_CHECKING:
    from techai_webutils.controllers.base import BaseController
    from collections.abc import Awaitable, Callable


def adapt[Req, Resp](
    controller: BaseController,
    handler: Callable[[Req], Awaitable[Resp]],
    parse: Callable[[Any], Awaitable[Req]],
    status: int = 200,
) -> Callable[[Any], Awaitable[dict[str, Any]]]:
    """Bridge a typed handler to a framework request handler.

    Args:
        controller: BaseController for error/response serialization.
        handler: Typed handler ``(Req) -> Resp``.
        parse: Async function that extracts ``Req`` from the raw request.
        status: HTTP status code for successful responses.

    Returns:
        An async function ``(raw_request) -> dict`` suitable for framework routing.

    """

    async def framework_handler(raw_request: object) -> dict[str, Any]:
        """Parse, dispatch, and serialize one request, converting errors to responses."""
        try:
            req = await parse(raw_request)
            result = await handler(req)
            return controller.write_json(status, result)
        except Exception as exc:
            return controller.write_error(exc)

    return framework_handler


def adapt_no_content[Req, Resp](
    controller: BaseController,
    handler: Callable[[Req], Awaitable[Resp]],
    parse: Callable[[Any], Awaitable[Req]],
) -> Callable[[Any], Awaitable[dict[str, Any]]]:
    """Bridge a typed handler that returns no content (204).

    Same as ``adapt`` but always returns 204 No Content on success.
    """

    async def framework_handler(raw_request: object) -> dict[str, Any]:
        """Parse and dispatch one request, returning 204 or an error response."""
        try:
            req = await parse(raw_request)
            await handler(req)
            return controller.write_no_content()
        except Exception as exc:
            return controller.write_error(exc)

    return framework_handler
